Intermediate Ruby: OOP

ebeagusamuel opened the pull request:
# Milestone3- Game Logic
- In this milestone, we encapsulate the different parts of the game logic in different classes.

krys2fa committed:
Add player input requests and display_board method

ebeagusamuel committed:
Add display promts for users

ebeagusamuel committed:
Fixed rubocop errors

ebeagusamuel committed:
Add methods to get players input

krys2fa committed:
Add recursion to get_player2_symbol method

krys2fa committed:
Add game methods to main.rb file

krys2fa committed:
Merge changes to branch

ebeagusamuel committed:
Add methds to keep playing the game

ebeagusamuel committed:
Fix rubocop errors

ebeagusamuel committed:
Remove un-necessary comments

ebeagusamuel committed:
Add method to display draw or win message

krys2fa committed:
Add classes to encapsulate parts of the game

krys2fa committed:
Update classes with methods from main.rb

krys2fa committed:
Add designs

ebeagusamuel committed:
Organize the code base into diffrent classes
- The different methods were moved to their appropriate classes.
- We added a counter to count for the number of wins each player has made at the end of the game.

Dandush03 commented on bin/main.rb:
```diff
@@ -1,2 +1,79 @@
+
+require './lib/player.rb'
+require './lib/board.rb'
+require './lib/validators.rb'
```
- [ ] please use `require_relative` for files in your directory instead of `require` is a ruby convention.

Dandush03 commented on lib/board.rb:
```diff
@@ -0,0 +1,24 @@
+class Board
+  attr_accessor :board
```
- [ ] Avoid using `attr_accessor`
 > HINT: `display_board`

Dandush03 commented on lib/player.rb:
```diff
@@ -0,0 +1,33 @@
+class Player
+  attr_reader :name, :symbol
+  attr_accessor :num_of_wins
```
[__OPTIONAL__]
Make a function that will set this attribute so there will be no needs for `attr_accessor` at all

Dandush03 commented on lib/validators.rb:
```diff
@@ -0,0 +1,29 @@
+  end
+
+  def self.won?(board, player)
+    winning_combinations = [
+      [0, 1, 2],
+      [3, 4, 5],
+      [6, 7, 8],
+      [0, 3, 6],
+      [1, 4, 7],
+      [2, 5, 8],
+      [0, 4, 8],
+      [2, 4, 6]
+    ]
```
- [ ] Please make a loop to validate de winning proccess

Dandush03 replied:
# Keep up the good work

### Happy Coding :computer: 

### Daniel Laloush

Dandush03 reviewed (commented):
# Status: :red_circle: Required Changes :red_circle: 

## Great Job So Far!

### Below you'll see some changes that need to be done in order to get this project approved.

- [ ] At the end of the game, there is no explicit way to end it.

krys2fa committed:
Refactor get player methods

krys2fa commented:
Hello @Dandush03,
Thanks for the review.
We have made the following corrections:
- we instructed users on how to explicitly end the game because the code already does that.
-  we updated 'require' to 'require_relative' as requested in the review.

However,
- the loop for validation as requested exists on line 19 in the lib/validators.rb file
- attr_accessors were only used where necessary as instructed by the milestone 3 project requirements.

krys2fa committed:
Move game methods to game class

krys2fa committed:
Fix errors raised by rubocop

krys2fa committed:
Refactor codes in all files

uryelah commented:
Hello **Samuel Ebeagu & Chris Amanor**, I'm Sarah, your reviewer for this PR 👩‍🚀  

Good job with your game loop, it's very thorough!

There's still just a requirement missing though, check the comments below 🦾 

--------

## Changes Required 🏗️ 🐛 💡

- [ ] As requested by pathwright, there should be no prints outside bin/main.rb. The reason is that way you can keep the concerns of the code, UI and logic, separated and easier to test and debug. Please refactor your code so it fits this requirement.
![only_main_print](https://user-images.githubusercontent.com/16193212/82450237-d2b93300-9a82-11ea-901c-11607b963ea9.png)


- [ ] In lib/Validators#won? you are both checking for a winning combination and storing the winning_combinations array. Please remember to keep your code as modular as possible. The array could be stored as a class variable instead, for example.

-----

 **Kindly fix the pointed out issues before requesting the next code review**

------------

Cheers and Happy coding!👏👏👏

Code review by **Sarah Uryelah Silva Chamorro**

(Feel free to contact me on slack by the handle of  **sarah** if you have any additional question)

ebeagusamuel committed:
Implement changes specified by TSE
- Moved all the gets and puts staatement from the logic files into the main.rb file.
- Changed winning combination in validator.rb into a class variable.

ebeagusamuel committed:
Correct rubocop errors

euqueme commented on bin/main.rb:
```diff
@@ -1,2 +1,110 @@
+    puts "#{player_one.name}, it is your turn..."
+    player_one
+  end
+end
```
- [ ] DRY rule should be respected. Don't repeat yourself. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/dry.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- The code in the if and else is very similar I think this method belongs in your Player class and whenever you use it here you could do something like this:

../lib/player.rb file
```ruby
def player_turn(player_one, player_two, curr_board)
  count = curr_board.integer_count
  if count.even?
    player_two
  else
    player_one
  end
end
```
../bin/main.rb file
```ruby
puts "#{player_turn(player_one, player_two, curr_board).name}, it is your turn..."
```

euqueme reviewed (changes requested):
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @ebeagusamuel and @krys2fa :lollipop: :balloon: 

### Thank you for doing all the changes requested by previous reviewers, your project is amazing I especially like the fact that you request for names for the players and you also keep track of the matches this makes your game more customizable and stands out from the rest :video_game:  :star: you've also have used OOP rules in your code almost perfectly :ok_hand:  there are still a couple more changes to make and you'll be good to go :runner: :running_man: 

#### Behaviour

- [ ] Your project is letting the players have empty names, let's make a validation to avoid this or set a default name like 'Player1' and 'Player2' if no name is given

![image](https://user-images.githubusercontent.com/52209067/82606923-1c824600-9b75-11ea-80ec-b48ea6347ce5.png)

#### Implementation Details

- [ ] I've left one comment in your code where a method should be optimized according to the DRY rule :point_down: 

### That's it, team! :raised_hands: 

### Please :pray: do the changes :technologist: :technologist: and submit back for a code review when you are done :memo: 

Happy coding! :rocket: 

Any questions? please ask me on slack **@ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES :vertical_traffic_light:

krys2fa committed:
Move player_turn method to Player class and fix user being able to proceed without a name

krys2fa committed:
Fix errors raised by rubocop

krys2fa commented:
Thank @euqueme for the review. The changes have been implemented as requested.

suretrust reviewed (approved):
Hi team, 👋 

### Great job implementing the suggested changes. 🙌 Let's get this merged! 👍 

_Good luck with your future projects and happy coding!_ 🎉

krys2fa commented:
Thanks @suretrust
